internal/transport/transport_test.go

```diff
@@ -2399,3 +2400,10 @@ func (s) TestClientDecodeHeaderStatusErr(t *testing.T) {
 		})
 	}
 }
+
+func TestConnectionError_Unwrap(t *testing.T) {
```

Comment:
This doesn't test the actual usage the user (you) are expecting.  Is it possible to add a test in `/test` at the `grpc.Dial` level for this, to make sure we don't break it in the future?

Reply:
@dfawley Sorry to see that this PR was closed. I didn't keep a close eye on it after it was approved.  The test is for the actual problem. It probably affects several different use-cases.

Reply:
The test is performed in an internal package, and does not go through any public APIs.  So the problem is, if we stop returning `transport.ConnectionError` and do something different instead, whatever externally-visible use case(s) you have will be broken, but none of our tests will fail.  That makes this test much less useful than something in `grpc/test`.  Let me know if you need any help figuring out how to add a test in that package, as it can be tricky.

Reply:
@dfawley thanks for offering to help figuring out how to make a more encompassing test. Unfortunately, I'm in a situation where I don't have more time to spend. I created a workaround in our code when we got hit by the regression. Then,  as a courtesy to you (since I really appreciate this package), I provided a PR with a fix, and a unit test to verify that the fix does what it's supposed to do. I also explained exactly how to reproduce the use-case to get the error from `DialContext`. I'm afraid that's as a far as I'm able to go with this, at least for the time being.

Thanks again for an excellent package.

Reply:
> I created a workaround in our code when we got hit by the regression.

AIUI you were relying on undocumented / unintentional behavior.  This whole flow isn't something we recommend to users -- `WithBlock` / `WithReturnConnectionError` are not things we recommend using, and are not provided in most/any other languages' implementations of gRPC.  As such, it is not a priority for us to maintain it.  If we take this PR without an appropriate test, there's a reasonable chance it will break again in the future, and fixing it will be an extremely low priority for us, given the above.  With a test, there's a much better chance we would notice it and prevent the breakage in the first place.

If you're okay with this risk and the lack of real support of this feature, we can merge this as-is.  If you would like to discuss other ways of doing what you're attempting that are better supported, then we are happy to help with that as well.

Reply:
Thanks for offering to help but we have good reasons for the approach chosen and no desire to change it at this point. I'd like you to merge this as-is.